File: C++/PalindromePermutationII.cc
// Given a string s, return all the palindromic permutations (without duplicates) of it.
// Return an empty list if no palindromic permutation could be form.

// For example:

// Given s = "aabb", return ["abba", "baab"].

// Given s = "abc", return [].

class Solution {
    bool nextPermutation(string& seed) {
        if (seed.size() < 2) {
            return false;
        }

        int i = seed.size() - 2;
        while (i >= 0 && seed[i] >= seed[i + 1]) {
            i--;
        }
        if (i < 0) {
            return false;
        }

        int j = seed.size() - 1;
        while (j > i && seed[i] >= seed[j]) {
            j--;
        }

        swap(seed[i], seed[j]);
        reverse(seed.begin() + i + 1, seed.end());
        return true;

    }
public:
    vector<string> generatePalindromes(string s) {
        unordered_map<char, int> dict;
        string seed = "";
        for (char c : s) {
            if (dict.count(c)) {
                seed += c;
                dict.erase(c);
            } else {
                dict[c]++;
            }
        }
        vector<string> output;
        if (dict.size() > 1) {
            return output;
        }
        string mid = dict.empty() ? "" : string(1, dict.begin()->first);
        sort(seed.begin(), seed.end());
        do {
            string t = seed;
            reverse(t.begin(), t.end());
            output.push_back(seed + mid + t);
        } while (nextPermutation(seed));
        return output;
    }
};

File: Java/PalindromePermutationII.java
// Given a string s, return all the palindromic permutations (without duplicates) of it. Return an empty
// list if no palindromic permutation could be form.

// For example:

// Given s = "aabb", return ["abba", "baab"].

// Given s = "abc", return [].

public class PalindromePermutationII {
    public List<String> generatePalindromes(String s) {
        Map<Character, Integer> mp = new TreeMap<>();
        for (int i = 0; i < s.length(); i++) {
            Integer cnt = mp.get(s.charAt(i));
            if (cnt == null) {
                mp.put(s.charAt(i), 1);
            } else {
                mp.put(s.charAt(i), 1 + cnt);
            }
        }
        String single = "";
        StringBuilder chars = new StringBuilder();
        for (Map.Entry<Character, Integer> entry : mp.entrySet()) {
            char c = entry.getKey();
            int val = entry.getValue();
            if (val % 2 == 1) {
                if (!single.isEmpty()) {
                    return new ArrayList<>();
                }
                single += c;
            }
            while (val >= 2) {
                chars.append(c);
                val -= 2;
            }
        }
        List<String> out = new ArrayList<>();
        boolean[] used = new boolean[chars.length()];
        generatePalindromes(chars.toString(), used, new StringBuilder(), single, out);
        return out;
    }

    void generatePalindromes(String s, boolean[] used, StringBuilder sb, String single, List<String> out) {
        if (sb.length() == s.length()) {
            out.add(sb.toString()+single+sb.reverse().toString());
            sb.reverse();
            return;
        }
        for (int i = 0; i < s.length(); i++) {
            if (!used[i]) {
                used[i] = true;
                sb.append(s.charAt(i));
                generatePalindromes(s, used, sb, single, out);
                used[i] = false;
                sb.deleteCharAt(sb.length() - 1);
                while (i+1 < s.length() && s.charAt(i+1) == s.charAt(i)) {
                    i++;
                }
            }
        }
    }
}
